slurm/src/slurmcc/allocation.py

```diff
@@ -0,0 +1,260 @@
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License.
+#
+import logging
+from typing import Callable, Dict, List, Set, Tuple
+
+from hpc.autoscale import util as hpcutil
+from hpc.autoscale import clock
+from hpc.autoscale.node.bucket import NodeBucket
+from hpc.autoscale.node.node import Node
+from hpc.autoscale.node.nodemanager import NodeManager
+from hpc.autoscale.results import AllocationResult, BootupResult
+
+from . import AzureSlurmError
+from . import partition as partitionlib
+from . import util as slutil
+
+
+def resume(
+    config: Dict,
+    node_mgr: NodeManager,
+    node_list: List[str],
+    partitions: List[partitionlib.Partition],
+) -> BootupResult:
+    name_to_partition = {}
+    for partition in partitions:
+        for name in partition.all_nodes():
+            name_to_partition[name] = partition
+    existing_nodes_by_name = hpcutil.partition(node_mgr.get_nodes(), lambda n: n.name)
+
+    nodes = []
+    for name in node_list:
+        if name in existing_nodes_by_name:
+            logging.info(f"{name} already exists.")
+            continue
+
+        if name not in name_to_partition:
+            raise AzureSlurmError(
+                f"Unknown node name: {name}: {list(name_to_partition.keys())}"
+            )
+        partition = name_to_partition[name]
+        bucket = partition.bucket_for_node(name)
+
+        def name_hook(bucket: NodeBucket, index: int) -> str:
+            if index != 1:
+                raise RuntimeError(f"Unexpected index: {index}")
+            return name
+
+        node_mgr.set_node_name_hook(name_hook)
+        constraints = {"node.bucket_id": bucket.bucket_id, "exclusive": True}
+        if partition.is_hpc:
```

Comment:
I believe we need to discuss this logic further, in my testing  I am noticing that autoscaling is working fine for hpc partitions, but does not work for HTC partitions (submit 10's of jobs in a for loop), and in the logs i see a runtim error coming from line 57 below:

2023-03-24 08:33:21,885 DEBUG: Candidates for job None: [NodeBucket(htc, available=3, size=Standard_F2s_v2, id=bbf47f74-25d3-47c5-bbcd-804da690dd3b)]
2023-03-24 08:33:21,885 DEBUG: Candidates for job None: [NodeBucket(htc, available=2, size=Standard_F2s_v2, id=bbf47f74-25d3-47c5-bbcd-804da690dd3b)]
2023-03-24 08:33:21,886 DEBUG: Candidates for job None: [NodeBucket(htc, available=1, size=Standard_F2s_v2, id=bbf47f74-25d3-47c5-bbcd-804da690dd3b)]
2023-03-24 08:33:21,887 DEBUG: Candidates for job None: [NodeBucket(htc, available=0, size=Standard_F2s_v2, id=bbf47f74-25d3-47c5-bbcd-804da690dd3b)]
2023-03-24 08:33:21,887 DEBUG: Full stacktrace
Traceback (most recent call last):
  File "/opt/azurehpc/slurm/venv/lib64/python3.6/site-packages/hpc/autoscale/clilib.py", line 1909, in main
    args.func(**kwargs)
  File "/opt/azurehpc/slurm/venv/lib64/python3.6/site-packages/slurmcc/cli.py", line 55, in wrapped
    return function(*args, **kwargs)
  File "/opt/azurehpc/slurm/venv/lib64/python3.6/site-packages/slurmcc/cli.py", line 209, in resume
    bootup_result = allocation.resume(config, node_mgr, node_list, partitions)
  File "/opt/azurehpc/slurm/venv/lib64/python3.6/site-packages/slurmcc/allocation.py", line 57, in resume
    raise RuntimeError()
RuntimeError


This is affecting the resume call being completed successfully, as then consequently the get_acct_info.sh also starts failing.